Only wrap exceptions when necessary in ArC subclassing

gsmet opened the pull request:
This way, the proxy has an exception handling as similar as possible to
the original class.

Fixes #329.

My initial inclination was to add the test in `SubclassGeneratorTest` but I couldn't figure out how to properly initialize my bean (and ended up with a lot of boilerplate code reimplementing a lot of logic). So I ended up moving the test in arc-tests to have the test infrastructure handy.

gsmet committed:
Only wrap exceptions when necessary in ArC subclassing

This way, the proxy has an exception handling as similar as possible to
the original class.

Fixes #329.

mkouba reviewed (changes requested):
Otherwise looks good!

gsmet committed:
Use DotNames.create() instead of DotName.createSimple()

Also take into account Throwable, even if not critical - it works OK
without it.
